src/py/flwr/client/secure_aggregation/secaggplus_handler.py

```diff
@@ -0,0 +1,453 @@
+
+    client: Optional[Union[Client, NumPyClient]] = None
+
+
+class SecAggPlusHandler(SecureAggregationHandler):
+    """Message handler for the SecAgg+ protocol."""
+
+    _shared_state = SecAggPlusState()
+    _current_stage = STAGE_UNMASKING
+
+    def handle_secure_aggregation(
+        self, named_values: Dict[str, Value]
+    ) -> Dict[str, Value]:
+        """Handle incoming message and return results, following the SecAgg+ protocol.
+
+        Parameters
+        ----------
+        named_values : Dict[str, Value]
+            The named values retrieved from the SecureAggregation sub-message
+            of Task message in the server's TaskIns.
+
+        Returns
+        -------
+        Dict[str, Value]
+            The final/intermediate results of the SecAgg+ protocol.
+        """
+        if not isinstance(self, (Client, NumPyClient)):
+            raise TypeError(
+                "The subclass of SecAggPlusHandler must be "
+                "the subclass of Client or NumPyClient."
+            )
+
+        self._check_stage(named_values)
+        stage = str(named_values.pop("stage"))
+        self._current_stage = stage
+
+        self._check_named_values(named_values)
+        if stage == STAGE_SETUP:
+            self._shared_state = SecAggPlusState(client=self)
+            return _setup(self._shared_state, named_values)
+        if stage == STAGE_SHARE_KEYS:
+            return _share_keys(self._shared_state, named_values)
+        if stage == STAGE_COLLECT_MASKED_INPUT:
+            return _collect_masked_input(self._shared_state, named_values)
+        if stage == STAGE_UNMASKING:
+            return _unmasking(self._shared_state, named_values)
+        raise ValueError(f"Unknown secagg stage: {stage}")
+
+    def _check_stage(self, named_values: Dict[str, Value]) -> None:
+        """Check the validity of the next stage."""
+        # Check the existence of 'stage'
+        if "stage" not in named_values:
+            KeyError(
+                "The required key 'stage' is missing from the input `named_values`."
+            )
+
+        # Check the validity of the next stage
+        stage = str(named_values["stage"])
+        if stage == STAGE_SETUP:
+            if self._current_stage != STAGE_UNMASKING:
+                log(WARNING, "restart from setup stage")
+        # if stage is not "setup",
+        # the stage from `named_values` should be the expected next stage
+        else:
+            expected_next_stage = STAGES[
+                (STAGES.index(self._current_stage) + 1) % len(STAGES)
+            ]
+            if stage != expected_next_stage:
+                raise ValueError(
+                    "Abort secure aggregation: "
+                    f"expect {expected_next_stage} stage, but receive {stage} stage"
+                )
+
+    # pylint: disable-next=too-many-branches
+    def _check_named_values(self, named_values: Dict[str, Value]) -> None:
```

Comment:
This could be a function:
```suggestion
    def _check_named_values(self, stage: str, named_values: Dict[str, Value]) -> None:
```